### PR TITLE
AZ itemized deduction

### LINKED ISSUE #2527: Arizona Itemized Deduction
TAX-140 Line 43
https://azdor.gov/forms/individual/form-140-resident-personal-income-tax-form-calculating
YEAR 2022

### PR DESCRIPTION
**Pull request by phywzc**
### Summary
📝📄📐

This pull request adds support for the Arizona itemized deduction adjustment for state income taxes, which is a feature of the Arizona tax system that affects the state income tax liability of filers who itemize their deductions. It creates a new parameter file `limit_from_federal_schedule_a.yaml` and a new variable file `az_itemized_deduction_adjustments.py` to implement the adjustment logic based on the 2022 Form 140 Schedule A.

> _`Arizona` tax_
> _Adjusting itemized deductions_
> _New files in `YAML`_

### Walkthrough
*  Add a new variable and parameter for the Arizona itemized deduction adjustment for state income taxes ([link](https://github.com/PolicyEngine/policyengine-us/pull/2571/files?diff=unified&w=0#diff-3f35d4fdf97754efff8c3413f8e844b8e7806a9269a49cc7edf9d2d87bba1c08R1-R20), [link](https://github.com/PolicyEngine/policyengine-us/pull/2571/files?diff=unified&w=0#diff-dcce5eb47a1317d2c0cfc62b2f2f0bf335be40aa7a3d5af83a28b96f73d5cebcR1-R65))
* Add an empty line to the changelog entry file ([link](https://github.com/PolicyEngine/policyengine-us/pull/2571/files?diff=unified&w=0#diff-5885042762833a4aef51405a4036067e4e6aefa7a5091a2987d1d9d1fc2647e8R1))

Fixes #2527

### CODE REVIEW

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: add changelog

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_total_itemized_deduction_allowed_by_irs.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_total_itemized_deduction_allowed_by_irs.py:20]:
```suggestion
        return add(tax_unit, period, items)
```

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_total_itemized_deduction_allowed_by_irs.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_total_itemized_deduction_allowed_by_irs.py:21]:
```suggestion
```

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_total_itemized_deduction_allowed_by_irs.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_total_itemized_deduction_allowed_by_irs.py:7]:
```suggestion
    label = " Total federal itemized deductions allowed to be taken on federal return"
```

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_total_itemized_deduction_allowed_by_irs.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_total_itemized_deduction_allowed_by_irs.py:4]:
```suggestion
class az_temized_deduction(Variable):
```

File: changelog.yaml
Comment (PavelMakarchuk) [inline changelog.yaml:1]: delete that piece - we only enter the changelog_entry.yaml -> the date will also be auto populated after we merge

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: add changelog here

File: policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/itemized_deduction_limit.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/itemized_deduction_limit.yaml:28]: delete all extra lines on the bottom but one

File: policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/itemized_deduction_limit.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/itemized_deduction_limit.yaml:3]:
```suggestion
  label: Arizona itemized deduction salt adjustment
```

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:26]: we can't use if, else in this case. We need to use a where(

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:30]: same, no if -> where(

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:72]: more descriptive variable names

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction_irs.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction_irs.py:1]: we already have a variable for it `itemized_deductions_less_salt` if we use it without the state and local tax or `tax_unit_itemizes` if full ded.

File: policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/itemized_deduction_limit.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/itemized_deduction_limit.yaml:9]: specific part of the paragraph

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:35]: split the comment into mult. pieces

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:75]: make the variables more explanatory

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions_less_salt.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions_less_salt.py:4]: we can just use `itemized_deductions_less_salt`

File: policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/itemized_deduction_limit.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/itemized_deduction_limit.yaml:1]: dont need this either as it is a federal cap as well

**Comment by phywzc** [comment]
In this specific case, do I need to make titles in test yaml descriptive?

**Comment by PavelMakarchuk** [comment]
> In this specific case, do I need to make titles in test yaml descriptive?

its good to make them descriptive as then it easier to remember which edge cases were tested and to see if maybe another test would be useful etc.

**Comment by phywzc** [comment]
> 

Sure, thanks for your comments. I will change those titles.

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:23]: can you use `medical_expense_deduction` to get the medical_expense_deduction?

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:39]:
```suggestion
```

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:44]: we can actually model the credit - we can either make a placeholdr variable or model the credit under these sections:
https://casetext.com/statute/arizona-revised-statutes/title-43-taxation-of-income/chapter-10-individuals/article-5-credits/section-43-1088-effective-until-ninety-one-days-after-adjournment-credit-for-contribution-to-qualifying-charitable-organizations-definitions

File: medical_expense
Comment (PavelMakarchuk) [inline medical_expense:1]: delete?

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:47]: I will check for the SALT deduction as they usually create a circular reference and we exclude them

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:48]: need to a real_estate_taxes

**Comment by phywzc** [comment]
I modeled "az_charitable_deduction" based on "https://casetext.com/statute/arizona-revised-statutes/title-43-taxation-of-income/chapter-10-individuals/article-5-credits/section-43-1088-effective-until-ninety-one-days-after-adjournment-credit-for-contribution-to-qualifying-charitable-organizations-definitions".

Test files have not been developed yet due to some outstanding issues that need to be resolved first.

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (shuyu7zhan) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:12]: need page number for both forms

File: policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/charity/ceiling/qualifying_foster.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/charity/ceiling/qualifying_foster.yaml:1]: description off

File: policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/charity/ceiling/qualifying_organization.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/charity/ceiling/qualifying_organization.yaml:15]: need 2021 values

File: policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/charity/ceiling/qualifying_foster.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/az/tax/income/deductions/itemized/charity/ceiling/qualifying_foster.yaml:15]: need 2021 values

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:49]:
```suggestion
```

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:50]:
```suggestion
```

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:54]:
```suggestion
```

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/variables/gov/states/az/tax/income/credits/charitable_contribution/az_charitable_contributions_credit.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/az/tax/income/credits/charitable_contribution/az_charitable_contributions_credit.py:38]:
```suggestion
        divisor = where(separate, p.separate_divisor, 1)
        return capped_credit / divisor
```

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:38]: This is equivalent to the code you wrote, but is it correct? You already skipped the medical expense deduction in the for loop, I think you can just add the entirety of medical_expense
```suggestion
        az_medical_expense_deduction = np.abs(medical_expense - medical_expense_deduction)
```

File: policyengine_us/variables/household/expense/charitable/foster_care_charitable_organization.py
Comment (MaxGhenis) [inline policyengine_us/variables/household/expense/charitable/foster_care_charitable_organization.py:4]: Make this a float.

This needs to be in the Arizona folder because of the bolded:
>5. " Qualifying foster care charitable organization" means a qualifying charitable organization that is exempt from federal income taxation under section 501(c)(3) of the internal revenue code or is a designated community action agency that receives community services block grant program monies pursuant to 42 United States Code section 9901 and that each operating year provides services to **at least two hundred qualified individuals in this state and spends at least fifty percent of its budget on services to qualified individuals in this state.**
```suggestion
class az_charitable_contributions_to_qualifying_foster_care_organizations(Variable):
```

File: policyengine_us/variables/gov/states/az/tax/income/credits/charitable_contribution/az_charitable_contributions_credit.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/az/tax/income/credits/charitable_contribution/az_charitable_contributions_credit.py:25]: This is based only on charitable contributions to qualifying organizations as defined in [§ 43-1088(M)(4)](https://law.justia.com/codes/arizona/2022/title-43/section-43-1088/). Make a new variable for that similar to my suggestion for the foster care one.

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:43]: why is this commented

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:57]: filers can still take a deduction for the amount not claimed as a credit.

since the credit will always be more valuable, we can assume they take the credit for the maximum amount allowable, then the deduction on any remainder

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:38]: image

Arizona seems to exclude the AGI floor

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:53]: delete commented lines

File: policyengine_us/variables/gov/states/az/tax/income/credits/az_total_credit.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/az/tax/income/credits/az_total_credit.py:1]: i think we can delete this file

File: policyengine_us/tests/policy/baseline/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.yaml
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.yaml:1]: make a couple more tests with varying medical expenses and charitable deduction

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (MaxGhenis) [reply policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py]: Right, excluding the federal deduction and adding all expenses will reflect that

**Comment by MaxGhenis** [comment]
Please resolve outstanding comment threads if they're addressed

**Comment by phywzc** [comment]
> Please resolve outstanding comment threads if they're addressed

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:6]: delete

File: medical_expense
Comment (MaxGhenis) [inline medical_expense:1]: delete

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deduction.py:4]:
```suggestion
class az_itemized_deductions(Variable):
```

File: policyengine_us/parameters/gov/states/az/tax/income/credits/charitable_contribution/ceiling/qualifying_foster.yaml
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/az/tax/income/credits/charitable_contribution/ceiling/qualifying_foster.yaml:1]:
```suggestion
description: Arizona caps the credit for contributions to a qualifying foster care charitable organization at this amount.
```

File: policyengine_us/parameters/gov/states/az/tax/income/credits/charitable_contribution/ceiling/qualifying_foster.yaml
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/az/tax/income/credits/charitable_contribution/ceiling/qualifying_foster.yaml:3]:
```suggestion
  label: Arizona credit for contributions to foster organizations cap
```

File: policyengine_us/parameters/gov/states/az/tax/income/credits/charitable_contribution/ceiling/qualifying_organization.yaml
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/az/tax/income/credits/charitable_contribution/ceiling/qualifying_organization.yaml:3]:
```suggestion
  label: Arizona charitable contribution credit cap
```

File: policyengine_us/parameters/gov/states/az/tax/income/credits/charitable_contribution/ceiling/qualifying_organization.yaml
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/az/tax/income/credits/charitable_contribution/ceiling/qualifying_organization.yaml:1]:
```suggestion
description: Arizona caps its credit for contributions to Arizona-based organizations at this amount.
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py:38]: Is abs correct? The deduction should never exceed the expenses, but I think a more natural way to address the possibility would be capping the difference at zero with max_(0, ...)

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py:49]:
```suggestion
        charitable_deduction_after_credit = max_(
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
Good to merge after these

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py
Comment (phywzc) [reply policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py]: I have modified the code based on your comments, but I am not certain if it aligns with what you were referring to.

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py:33]:
```suggestion
        # Arizona allows a deduction for medical and dental expenses
        # that cannot be deducted from federal income taxes.
```

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py:44]: Please keep comments to 79 characters unless impossible (e.g. due to a long URL)
```suggestion
        # Amount of charitable contributions for which you are claiming 
        # a credit under Arizona law.
```

**Review by MaxGhenis — APPROVED** [review]

File: policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py
Comment (MaxGhenis) [reply policyengine_us/variables/gov/states/az/tax/income/deductions/itemized/az_itemized_deductions.py]: yep, thanks!

**Pull request merged**
